server/src/frameworks/Foundry/FoundryProject.ts

```diff
@@ -0,0 +1,286 @@
+/* eslint-disable @typescript-eslint/no-explicit-any */
+import { analyze } from "@nomicfoundation/solidity-analyzer";
+import fs from "fs";
+import { CompilerInput } from "hardhat/types";
+import _ from "lodash";
+import path from "path";
+import semver from "semver";
+import { DidChangeWatchedFilesParams } from "vscode-languageserver-protocol";
+import { analyzeSolFile } from "../../parser/analyzer/analyzeSolFile";
+import { OpenDocuments, ServerState } from "../../types";
+import { toUnixStyle } from "../../utils";
+import directoryContains from "../../utils/directoryContains";
+import { getOrInitialiseSolFileEntry } from "../../utils/getOrInitialiseSolFileEntry";
+import { runCmd } from "../../utils/operatingSystem";
+import CompilationDetails from "../base/CompilationDetails";
+import Project from "../base/Project";
+import { Remapping } from "./Remapping";
+
+interface SolFileDetails {
+  path: string;
+  pragmas: string[];
+}
+
+export default class FoundryProject extends Project {
+  public priority = 1;
+  public sourcesPath!: string;
+  public testsPath!: string;
+  public remappings: Remapping[] = [];
+  public initializeError?: string;
+  public configSolcVersion?: string;
+
+  constructor(
+    serverState: ServerState,
+    basePath: string,
+    public configPath: string
+  ) {
+    super(serverState, basePath);
+  }
+
+  public id(): string {
+    return this.configPath;
+  }
+
+  public frameworkName(): string {
+    return "Foundry";
+  }
+
+  public async initialize(): Promise<void> {
+    try {
+      const config = JSON.parse(
+        await runCmd("~/.foundry/bin/forge config --json", this.basePath)
+      );
+      this.sourcesPath = path.join(this.basePath, config.src);
+      this.testsPath = path.join(this.basePath, config.test);
+      this.configSolcVersion = config.solc || undefined; // may come as null otherwise
+
+      const rawRemappings = await runCmd(
+        "~/.foundry/bin/forge remappings",
+        this.basePath
+      );
+      this.remappings = this._parseRemappings(rawRemappings);
+    } catch (error: any) {
+      this.serverState.logger.error(error.toString());
+
+      switch (error.code) {
+        case 127:
+          this.initializeError =
```

Comment:
This has been my reasoning but I'm open to hear an alternative approach: I think it needs to be stored because initialization and validation are different invocations. On each validation, if the initialization was successful, a failure message is returned (and then a status item in the UI). I think the "interpretation" of the execution error (ExecError => string error reason) is ok to be in this place since the caller doesn't have to know what's going under the hood (invoking forge through shell).